cobalt/migrate_storage_record/migrate.cc

```diff
@@ -0,0 +1,398 @@
+      initial_url = url;
+    } else {
+      LOG(ERROR) << "URL \"" << url_switch
+                 << "\" from parameter is not valid, using default URL "
+                 << initial_url;
+    }
+  }
+
+  return initial_url;
+}
+
+std::string GetPartitionKey() {
+  auto* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(browser::switches::kLocalStoragePartitionUrl)) {
+    std::string local_storage_partition_url = command_line->GetSwitchValueASCII(
+        browser::switches::kLocalStoragePartitionUrl);
+    auto partition_key = GetApplicationKey(GURL(local_storage_partition_url));
+    CHECK(!partition_key.empty())
+        << "local_storage_partition_url is not a valid URL.";
+    return partition_key;
+  }
+  return GetApplicationKey(GetInitialURL());
+}
+
+bool ShouldFallbackToDefaultId(const std::string& key) {
+  return key == GetApplicationKey(GURL(kDefaultURL));
+}
+
+std::unique_ptr<cobalt::storage::Storage> ReadStorage(const GURL& url) {
+  std::string encoded_url = GetApplicationKey(url);
+  if (encoded_url.empty()) {
+    LOG(ERROR) << "Invalid url: " << url;
+    return nullptr;
+  }
+  auto record = std::make_unique<starboard::StorageRecord>(encoded_url.c_str());
+  if (!record->IsValid() || record->GetSize() <= 4) {
+    record->Delete();
+    LOG(ERROR) << "No valid storage record found.";
+    if (!ShouldFallbackToDefaultId(encoded_url)) {
+      return nullptr;
+    }
+    LOG(INFO) << "Trying fallback storage record.";
+    record = std::make_unique<starboard::StorageRecord>();
+    if (!record->IsValid() || record->GetSize() <= 4) {
+      record->Delete();
+      LOG(ERROR) << "No valid fallback storage record found.";
+      return nullptr;
+    }
+  }
+  base::ScopedClosureRunner delete_record(base::BindOnce(
+      [](starboard::StorageRecord* record) {
+        if (record && record->IsValid()) {
+          record->Delete();
+        }
+      },
+      record.get()));
+  auto bytes = std::vector<uint8_t>(record->GetSize());
+  int read_result =
+      record->Read(reinterpret_cast<char*>(bytes.data()), bytes.size());
+  if (read_result != bytes.size()) {
+    LOG(WARNING) << "Storage record size mismatch " << read_result << " "
+                 << bytes.size();
+    return nullptr;
+  }
+
+  std::string version(reinterpret_cast<const char*>(bytes.data()), 4);
+  if (version != "SAV1") {
+    return nullptr;
+  }
+
+  auto storage_data = std::make_unique<cobalt::storage::Storage>();
+  if (!storage_data->ParseFromArray(
+          reinterpret_cast<const char*>(bytes.data() + 4), bytes.size() - 4)) {
+    LOG(ERROR) << "Unable to parse storage with size " << bytes.size();
+    return nullptr;
+  }
+  return storage_data;
+}
+
+void SetCanonicalCookieCallback(std::atomic_int* pending,
+                                base::OnceClosure* all_done_callback,
+                                net::CookieAccessResult set_cookie_result) {
+  --(*pending);
+  if (pending->load() == 0) {
```

Comment:
nit: Could we just do `if (--(*pending) ==0 ) {`?

std::atomic::operator-- returns the decreased value (https://en.cppreference.com/w/cpp/atomic/atomic/operator_arith).

Same in l.279, maybe others.